### PR TITLE
Cache redesign part 2

### LINKED ISSUE #740: Cache redesign part 2: Send every request
## References
This ticket corresponds to issue 7 from #718. 

That ticket was too big to tackle at once so we split it up in to smaller, more manageable chunks of work.

## Description
Our attempts at abstracting away the cache has made working with it harder not easier. 
Every time any component needs anything from the rest api, they call the dataservice, which creates a request object, and passes it off to the request service. The request service then checks if the thing being requested is already cached, if it is it doesn't actually send the request, but returns the cached result. 

The original reason to do this was that developers should just program as if there was no cache and we'd be able to disable the cache or swap it out with a different implementation etc. 

In practice it doesn't work like that at all. Every component or service that interacts with the request service has to take in to account this behavior, there are cases where we can't help but force bypass this behavior, it has led to complicated code with a bunch of edge cases, a bug chunk of memory that's needed to track all the requests that were never sent etc. We're not even close to being able to turn off the cache without breaking absolutely everything.

## Proposed solution

- The requestservice will send every request it receives
- dataservices will check the cache for a valid object before sending a request
- Only if the check fails (the object isn't cached or is stale), the dataservice will send a request to the request service.
- If a certain method for a specific service *never* needs a cached object, it can simply leave out the check.
- We no longer need forceBypassChace, the request id mapping index, or observable.race() calls for retrieving a request which should make everything more straightforward and should improve performance.

This will take an estimated **35 - 45h**

### PR DESCRIPTION
**Pull request by artlowel**
## References
* Depends on #961
* Fixes #740

## Description
This PR moves the responsibility of deciding when to use a cached version of a request out of the RequestService, and to the dataservices and components.

`RequestService.configure(request)` has been replaced by `RequestService.send(request, useCached = false)`. Any component or service that calls it can decide whether or not a cached version is applicable for the use case. 

Requests that don't get sent because a cached version was available will no longer be indexed in the store and redirected to the cached request. This removes a lot of complexity from the methods that retrieve objects from the store, and frees client-side memory that was used to store all those unsent requests.

In order to reduce code duplication, as many dataservices methods as possible have been refactored to defer to findByHref instead of constructing their own GET requests, sending them, and building the RemoteData for the response. 

A single HrefOnlyDataService was also created to replace all dataservices that only needed findByHref methods, and nothing custom. You can now simply add another`@dataService(type)` annotation to the HrefDataService if you have a type for which this is the case.

## Instructions for Reviewers
The way to test this PR is, unfortunately, to try everything, as this PR affects anything that uses data from the rest api. 

Keep an eye on the requests made in the network tab of your browser's dev tools to ensure that requests that should be cached aren't needlessly re-requested. 

In order to see what happens when objects become stale and _do_ need to be re-requested, it's useful to reduce the default cache duration in your environment file. This will set it to 15 seconds:
```typescript
  cache: {
    msToLive: {
      default: 15 * 1000
    }
  }
  ```

Before reporting an issue, please verify that it doesn't happen on https://dspace7-demo.atmire.com

## Checklist
- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

File: src/app/core/data/href-only-data.service.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+@dataService(VOCABULARY_ENTRY)
+@dataService(ITEM_TYPE)
+@dataService(LICENSE)
+export class HrefOnlyDataService {
```
Comment (tdonohue) [inline src/app/core/data/href-only-data.service.ts:47]: Please add some TypeDocs to describe this new service.

Also, this new service doesn't appear to have any specs?

File: src/app/core/data/request.service.spec.ts [hunk 2]
```diff
@@ -198,44 +180,7 @@ describe('RequestService', () => {
       });
     });
 
-    describe(`if the request with the specified UUID wasn't sent, because it was already cached`, () => {
```
Comment (tdonohue) [inline src/app/core/data/request.service.spec.ts:201]: Is there a reason we removed the test here to verify we are obtaining the cached version of the request?  Is this now tested elsewhere?

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts [hunk 3]
```diff
@@ -161,11 +161,13 @@ export class RelationshipEffects {
   }
 
   private addRelationship(item1: Item, item2: Item, relationshipType: string, submissionId: string, nameVariant?: string) {
+    console.log('addRelationship 1');
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts:164]: This looks to be accidental & can be removed

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts [hunk 4]
```diff
@@ -161,11 +161,13 @@ export class RelationshipEffects {
...
     const type1: string = item1.firstMetadataValue('relationship.type');
     const type2: string = item2.firstMetadataValue('relationship.type');
     return this.relationshipTypeService.getRelationshipTypeByLabelAndTypes(relationshipType, type1, type2)
       .pipe(
         mergeMap((type: RelationshipType) => {
+            console.log('addRelationship 2', type.id, item2, item1, nameVariant);
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts:170]: Same here... This looks to be accidental & can be removed

File: src/app/shared/utils/follow-link-config.model.ts [hunk 5]
```diff
@@ -28,6 +28,17 @@ export class FollowLinkConfig<R extends HALResource> {
...
+  /**
+   * If this is true, the link will only be retrieved if there's no valid cached version. Defaults
+   * to true
+   */
+  useCachedVersionIfAvailable? = true;
+
+  /**
+   *
+   */
+  reRequestOnStale? = true;
```
Comment (tdonohue) [inline src/app/shared/utils/follow-link-config.model.ts:41]: Please fill out the (currently empty) description above.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@artlowel : I gave this a review & test today.  First off, I found some minor things in the code (see inline below).  In tests, things _mostly work well_.  

However, I've been able to find a few areas of the application where I seemingly hit an infinite loop of requests (the app hangs & in Chrome DevTools the number of requests increases so rapidly that you cannot even see the new requests in the Network tab's table).  I've tried reproducing the same behavior on our demo site & haven't been able to do so. 

Here's three ways I've hit this infinite loop (There may be more that I haven't discovered yet, as I haven't clicked every button. But, maybe you'll notice a similarity between these three):
1. Go to the Groups page (http://localhost:4000/admin/access-control/groups), Edit a Group, and add a new EPerson.
2. Go to the Groups page (http://localhost:4000/admin/access-control/groups), Edit a Group, and **remove** an EPerson.
3. Go to the Processes page  (http://localhost:4000/processes), click on a previously run Process and click on the "Retrieve process output" button

File: src/app/core/data/href-only-data.service.ts (hunk 1 above)
Comment (artlowel) [reply src/app/core/data/href-only-data.service.ts:54]: Added docs and specs

File: src/app/core/data/request.service.spec.ts (hunk 2 above)
Comment (artlowel) [reply src/app/core/data/request.service.spec.ts:212]: The behavior that test was verifying has been removed by this PR

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts (hunk 3 above)
Comment (artlowel) [reply src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts]: Done

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts (hunk 4 above)
Comment (artlowel) [reply src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.ts]: Done

File: src/app/shared/utils/follow-link-config.model.ts (hunk 5 above)
Comment (artlowel) [reply src/app/shared/utils/follow-link-config.model.ts:42]: Done

**Comment by artlowel** [comment]
@tdonohue and @atarix83 This PR has been rebased on main, and all issues reported so far have been fixed. It's ready to be reviewed once more.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@artlowel : Thanks for the updates!  It's looking better overall...all my prior feedback was fixed.  Additionally, overall the application seems slightly quicker. However, I've found a few more odd behaviors in different areas this time. (I'll note, I'm not able to easily verify whether these exist on the demo site as the demo site won't let me login)

1. In the Metadata Registry, if I edit the "dc" registry and **add** or **remove** a field, the action works, but I notice in Chrome DevTools that *every* schema seems to be reloaded from the REST API (via an `OPTIONS` and `GET` request, totally ~50 requests to the REST API). 
2. On the Processes page, if I click on a previously run process & then click the "Output Files" link, nothing happens.  I'd expect the file to be downloaded.  However, the "Retrieve process output" button now works.  (This might be an existing bug, if so, I can log a separate ticket)
3. Creating a new EPerson from the EPerson screen seems to work, but I never get sent back to the list of EPersons.  I have to manually reload the page to get back there.  (This also might be an existing bug, if so I can log a separate ticket)

Overall though, I'm not finding any major issues. Everything above is relatively minor bugs.  So, overall, I'm in favor of moving this forward quickly.

**Comment by artlowel** [comment]
>I'm not able to easily verify whether these exist on the demo site as the demo site won't let me login

https://demo7.dspace.org/ is live already, you can use that to test. It should be fully functional. The only thing still missing is a redirect from dspace7-demo.atmire.com to demo7.dspace.org. I can't do that myself. But it should be done in the next few days.

>1. In the Metadata Registry, if I edit the "dc" registry and add or remove a field, the action works, but I notice in Chrome DevTools that every schema seems to be reloaded from the REST API (via an OPTIONS and GET request, totalling ~50 requests to the REST API).

That was a pre-existing issue, but I saw an easy way to fix it, so I did.

>2. On the Processes page, if I click on a previously run process & then click the "Output Files" link, nothing happens. I'd expect the file to be downloaded. However, the "Retrieve process output" button now works. (This might be an existing bug, if so, I can log a separate ticket)

That did get accidentally broken by this PR. Fixed.

>3. Creating a new EPerson from the EPerson screen seems to work, but I never get sent back to the list of EPersons. I have to manually reload the page to get back there. (This also might be an existing bug, if so I can log a separate ticket)

That was also a pre-existing issue. Again, easy enough to fix. So it's fixed as well.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @artlowel .  This is working great for me.  All my prior feedback & reported bugs have been resolved.  So, I feel this is good enough to merge!

@atarix83 : If you have the chance could you prioritize a review of this PR?  Several other PRs are dependent on it.  Thanks!

**Review by atarix83 — COMMENTED** [review]
thanks @artlowel for this important work.

I've reviewed the code and all looks good to me.

While testing I've found some minor issues: 

- in the edit item metadata page (e.g. http://localhost:4000/items/6c33230e-5d6a-4f98-b4e7-3a5c39982897/edit/metadata) if you edit the value of an existing metadata, save and then move to another tab (status or any other tab) when you get back to the metadata tab you have the old value for the metadata changed previously
- if you edit any resource policy (http://localhost:4000/items/6c33230e-5d6a-4f98-b4e7-3a5c39982897/edit/authorizations/edit?policyId=769) the field that show eperson or group related to the policy is empty 
Schermata da 2021-02-04 10-55-28

**Comment by tdonohue** [comment]
@artlowel : Apologies, it looks like I created a minor merge conflict in this PR by merging #968 .  I'll refrain from merging anything else on the Angular side until this one is ready to go, as I'd like to get this merged as soon as reasonably possible.

**Comment by artlowel** [comment]
@atarix83 Thanks for your review
Both issues have been fixed

**Review by tdonohue — APPROVED** [review]
👍 This looks good.  I've tested the minor bugs that @atarix83 found, and I've verified they are now resolved.

Because @atarix83 [said previously](https://github.com/DSpace/dspace-angular/pull/975/#pullrequestreview-583218520) 
> I've reviewed the code and all looks good to me.

I'm going to consider that a +1 from him, now that the minor bugs are fixed.  Therefore, I'll merge this immediately as it has had two approving reviewers. 

Normally, I'd wait for @atarix83 to take a second quick look, but I'm aware of how many other PRs have become dependent on this one & are currently "blocked".  That said, @atarix83 if you notice anything else that needs immediate resolution in this PR, please do let me know & I'll make it high priority to get resolved in a follow-up PR.

**Pull request merged**